### ISSUE TITLE
[no-export] feature request: allow contract testing pattern

### ISSUE THREAD

**Issue by davidje13** [jest-community/eslint-plugin-jest#666]
`jest/no-export` should not produce errors unless `describe` or `it` blocks appear in the global scope for a file. A use-case for this is contract testing:

### commutative.contract.js

```
export default (functionUnderTest) => {
  it('gives the same answer regardless of parameter order', () => {
    const a = Math.round(Math.random() * 1000);
    const b = Math.round(Math.random() * 1000);
    expect(functionUnderTest(a, b)).toEqual(functionUnderTest(b, a));
  });
};
```

### add.test.js

```
import checkCommutative from './commutative.contract.ts';

describe('add', () => {
  checkCommutative(add);
});
```

### multiply.test.js

```
import checkCommutative from './commutative.contract.ts';

describe('multiply', () => {
  checkCommutative(multiply);
});
```

Currently `jest/no-export` flags the contract file as it contains both `it` and `export`.

**Comment by G-Rath** [comment]
For this pattern you should disable `no-export`, as it's exactly what that rule is meant to lint for.

You can use [overrides](https://eslint.org/docs/user-guide/configuring#disabling-rules-only-for-a-group-of-files) to disable the rule for specific files.

**Comment by davidje13** [comment]
@G-Rath yes, that's what I currently do, but it would be nice to have it supported directly. As I understand it, the rule is intended to avoid situations like this:

### add.test.js

```javascript
export const exampleNumber = () => 7;

it('adds numbers', () => {
  expect(add(exampleNumber, 2)).toEqual(9);
});
```

### subtract.test.js

```javascript
import { exampleNumber } from './add.test.ts'; // oops! this will run the add tests a second time!

// ...
```

i.e. situations where just importing a file will cause tests to run. This feels quite different from the example I'm talking about where the import causes no tests to run.

I can't think of a file which does not contain `describe` or `it` in the global scope but should still trigger this error. Of course it might be possible that somebody uses an IIFE in their file which would probably make it difficult/impossible to realise the tests will run just by importing the file, but that's not a pattern I've ever actually seen, even in bad code.
